### PR TITLE
IGNITE-14607: Regex Based Filtering For IP Addresses During Discovery

### PR DESCRIPTION
**Pull request by atris**
This commit introduces functionality to filter IP addresses during IP discovery using regexes.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import org.apache.ignite.spi.IgniteSpiException;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.junit.Test;
+
+/**
+ * Test discovery SPI with address filter present
+ */
+public class TcpDiscoveryWithAddressFilterTest extends TcpDiscoveryWithWrongServerTest {
+
+    IgnitePredicate<InetSocketAddress> addressFilter = new P1<InetSocketAddress>() {
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java:37]: Javadoc is needed for all fields, please also specify access modifier.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        startTcpThread(new NoResponseWorker(), SERVER_PORT);
+        startTcpThread(new NoResponseWorker(), LAST_SERVER_PORT);
+
+        simpleTest();
+    }
+
+    /**
+     * Check for incoming addresses and check that the filter was applied
+     */
+    class TcpDiscoveryWithAddressFilter extends TcpDiscoverySpi {
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java:78]: please use protected or private modifier

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithWrongServerTest.java [hunk 3]
```diff
@@ -43,16 +43,16 @@
  */
 public class TcpDiscoveryWithWrongServerTest extends GridCommonAbstractTest {
     /** Non-Ignite Server port #1. */
-    private static final int SERVER_PORT = 47500;
+    static final int SERVER_PORT = 47500;
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithWrongServerTest.java:46]: please use protected here & for all fields, classes below

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        /** {@inheritDoc} */
+        @Override protected Collection<InetSocketAddress> resolvedAddresses() throws IgniteSpiException {
+            Collection<InetSocketAddress> res = super.resolvedAddresses();
+
+            for (InetSocketAddress addr : res) {
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java:83]: curly braces not needed for single-statement, one-line body.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import org.apache.ignite.internal.util.typedef.P1;
+import org.apache.ignite.lang.IgnitePredicate;
+import org.apache.ignite.spi.IgniteSpiException;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.junit.Test;
+
+/**
+ * Test discovery SPI with address filter present
+ */
+public class TcpDiscoveryWithAddressFilterTest extends TcpDiscoveryWithWrongServerTest {
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryWithAddressFilterTest.java:35]: You need to add this class to some test suite.

**Review by alamar — CHANGES_REQUESTED** [review]

**Comment by atris** [comment]
@alamar Updated, please see.

**Review by vkulichenko — APPROVED** [review]

**Pull request closed without merging**